### PR TITLE
Use abseil to print stacktrace when signal is raised in ChainerX

### PR DESCRIPTION
**Pull request by take-cheeze**
rel: #7420
Uses [failure_signal_handler.h](https://github.com/abseil/abseil-cpp/blob/38b704384cd2f17590b3922b97744be0b43622c9/absl/debugging/failure_signal_handler.h) to print stacktrace to stderr.
Only enabled in debug built ChainerX.

Example output in my working buggy code:
```
tests/chainer_tests/functions_tests/connection_tests/test_convolution_2d.py ......python3.7: ~/dev/chainer/chainerx_cc/chainerx/stack_vector.h:146: chainerx::StackVector<T, N>::iterator chainerx::StackVector<T, N>::emplace(chainerx::StackVector<T, N>::const_itera\
tor, Args&& ...) [with Args = {signed char&}; T = signed char; long unsigned int N = 10; chainerx::StackVector<T, N>::iterator = signed char*; chainerx::StackVector<T, N>::const_iterator = const signed char*]: Assertion `n_ < N' failed.
*** SIGABRT received at time=1560405511 ***
PC: @     0x7fdd103d1077  (unknown)  (unknown)
    @     0x7fdce34e4fdb         64  absl::lts_2018_12_18::WriteFailureInfo()
    @     0x7fdce34e51c3         96  absl::lts_2018_12_18::AbslFailureSignalHandler()
    @     0x7fdd10724dd0  816254376  (unknown)
    @ 0x4173257325203a75  (unknown)  (unknown)
```

Things need to be discussed:
- [x] Whether Apache 2 license of abseil is OK
- [x] Should be enabled in release build too and switch with environment variable -> https://github.com/chainer/chainer/projects/2#card-22844563
- [x] When to init stacktracer(currently in python module initializer) -> https://github.com/chainer/chainer/projects/2#card-22844614
- [x] Whether to replace other third_party libraries(gsl::span, optional) with abseil -> https://github.com/chainer/chainer/projects/2#card-22844632 https://github.com/chainer/chainer/projects/2#card-22844630

### CODE REVIEW

**Comment by niboshi** [comment]
Jenkins, test this please

**Comment by niboshi** [comment]
It seems we have to disable `-Werror` for building abseil.
```
2019-06-17 12:33:15.761457 STDOUT 2565] | [ 62%] Building CXX object chainerx/abseil-build/absl/strings/CMakeFiles/str_format_internal.dir/internal/str_format/parser.o
-- | --
2019-06-17 12:33:15.878856 STDERR 2565] | /root/chainerx_build/chainerx/abseil/absl/time/internal/cctz/src/time_zone_info.cc: In member function 'virtual bool absl::lts_2
2019-06-17 12:33:15.878989 STDERR 2565] | 018_12_18::time_internal::cctz::TimeZoneInfo::NextTransition(absl::lts_2018_12_18::time_internal::cctz::time_point<std::chrono::
2019-06-17 12:33:15.879006 STDERR 2565] | duration<long int> >&, absl::lts_2018_12_18::time_internal::cctz::time_zone::civil_transition*) const':
2019-06-17 12:33:15.879109 STDERR 2565] | /root/chainerx_build/chainerx/abseil/absl/time/internal/cctz/src/time_zone_info.cc:925:41: error: missing initializer for member
2019-06-17 12:33:15.879120 STDERR 2565] | 'absl::lts_2018_12_18::time_internal::cctz::Transition::type_index' [-Werror=missing-field-initializers]

2019-06-17 12:33:15.879132 STDERR 2565] | const Transition target = { unix_time };
-- | --
2019-06-17 12:33:15.879144 STDERR 2565] | ^
 
```

**Comment by niboshi** [comment]
>- [ ] Whether Apache 2 license of abseil is OK

OK.

> - [ ] Should be enabled in release build too and switch with environment variable
> - [ ] When to init stacktracer(currently in python module initializer)
> - [ ] Whether to replace other third_party libraries(gsl::span, optional) with abseil

These can be done later.
I have created respective cards on the project board.

**Comment by take-cheeze** [comment]
Made a workaround to disable `-Werror` flag in abseil build.
Hope it works.

**Comment by niboshi** [comment]
Hmm it looks tricky.
How about using `set_source_files_properties`? https://stackoverflow.com/a/13639476/402996

**Comment by take-cheeze** [comment]
Seems the warning is fixed in newer abseil: https://github.com/abseil/abseil-cpp/issues/261

**Comment by niboshi** [comment]
Oh, I see.
I still think it's better to disable `-Werror` for external libs, but perhaps in other chance?

File: chainerx_cc/CMakeLists.txt [hunk 1]
```diff
@@ -163,10 +163,11 @@ if("${CMAKE_CXX_COMPILER_ID}" MATCHES "GNU|Clang|Intel")
     # cmake -DCHAINERX_WARNINGS_AS_ERRORS=ON|OFF
     if (CHAINERX_WARNINGS_AS_ERRORS)
         if (MSVC)
-            set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} /W4")
+            set(CHAINERX_WARNINGS_AS_ERRORS_FLAGS "${CMAKE_CXX_FLAGS} /W4")
```
Comment (niboshi) [inline chainerx_cc/CMakeLists.txt]: This fix can be reverted as well.

File: chainerx_cc/chainerx/third_party/abseil.cmake [hunk 2]
```diff
@@ -0,0 +1,14 @@
+cmake_minimum_required(VERSION 3.1)
+project(abseil-download NONE)
+
+include(ExternalProject)
+ExternalProject_Add(pybind11
+    GIT_REPOSITORY    https://github.com/abseil/abseil-cpp.git
+    GIT_TAG           72e09a54d993b192db32be14c65adf7e9bd08c31
```
Comment (niboshi) [inline chainerx_cc/chainerx/third_party/abseil.cmake]: In principal we should not rely on a pre-release commit.
How about adding a TODO?

**Comment by take-cheeze** [comment]
BTW I tried `set_source_files_properties` but it didn't work well with subdirectory project.
CMake is difficult tool...

**Comment by niboshi** [comment]
@take-cheeze 
Hmm, from documentation it looks it should work with subdirectories. :thinking: 
Anyway thank you for information!

Jenkins, test this please

**Pull request merged**